### PR TITLE
fix(ui): command line issues with 'cmdheight' set to zero

### LINKED ISSUE #20888: multiple statuslines with external commandline when resizing windows
### Neovim version (nvim -v)

NVIM v0.9.0-dev-77-gf175ca9f7

### Vim (not Nvim) behaves the same?

not relevant I think

### Operating system/version

macOS 11.4

### Terminal name/version

kitty 0.24.4

### $TERM environment variable

xterm-kitty

### Installation

Download releases

### How to reproduce the issue

I'm reproducing this issue with `noice.nvim` because just writing something with external commandline that doesn't make neovim completely unusable is quite hard.
```lua
-- minimal.lua
vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvim/site]])
local package_root = "/tmp/nvim/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
    require("packer").startup({
        {
            "wbthomason/packer.nvim",
            {
                "folke/noice.nvim",
                requires = {
                    "MunifTanjim/nui.nvim",
                    "rcarriga/nvim-notify",
                },
            },
        },
        config = {
            package_root = package_root,
            compile_path = install_path .. "/plugin/packer_compiled.lua",
            display = { non_interactive = true },
        },
    })
end
_G.load_config = function()
    vim.o.lazyredraw = false
    require("noice").setup({})
end
if vim.fn.isdirectory(install_path) == 0 then
    print("Installing plugins and dependencies.")
    vim.fn.system({
        "git",
        "clone",
        "--depth=1",
        "https://github.com/wbthomason/packer.nvim",
        install_path,
    })
end
load_plugins()
require("packer").sync()
vim.cmd(
    [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
)
```
1. `nvim --clean -u minimal.lua`
2. `2<c-w>-`
3. `<c-w>+`
4. `<c-w>+`

### Expected behavior

statusline goes all the way down again

### Actual behavior

two statuslines above each other

### LINKED ISSUE #21558: cmdheight=0 works weirdly with `resize`
### Neovim version (nvim -v)

v0.9.0-dev-565+gd215dae0e

### Vim (not Nvim) behaves the same?

no

### Operating system/version

linux 6.1.1

### Terminal name/version

kitty 0.26.5

### $TERM environment variable

xterm-kitty

### Installation

Build from source

### How to reproduce the issue

nvim --clean
:set cmdheight=0
:resize -1
:resize +1

### Expected behavior

As it's still experimental, I don't know what should the expected behavior be
For me, it's that cmdheight should stay as 0, or to be back to 0 after resize +1

### Actual behavior

cmdheight is now back to 1, and have to be reset to 0

### PR DESCRIPTION
**Pull request by luukvbaal**
**fix(ui): allow resize commands to set 'cmdheight' to 0**

Resolve https://github.com/neovim/neovim/issues/21558

**fix: issues with command line if ui elements are externalized**

Resolve https://github.com/neovim/neovim/issues/20888 and handle side effects for setting `'cmdheight'` to zero.

### CODE REVIEW

**Comment by max397574** [comment]
would this perhaps also fix this issue https://github.com/neovim/neovim/issues/20888 ?

**Comment by justinmk** [comment]
does this need a test? (or just add a small tweak to an existing test)

**Comment by luukvbaal** [comment]
> would this perhaps also fix this issue https://github.com/neovim/neovim/issues/20888 ?

I don't think so.

> does this need a test? (or just add an extra case to an existing test)

Sure, but I think we first need to decide if it's actually supposed to work as proposed. I think it makes sense to allow resizing the cmdline height back to 0 if it was set explicitly but could be debatable.

I'll look into the issue and testcase this weekend.

File: test/functional/ui/cmdline_spec.lua [hunk 1]
```diff
@@ -1364,4 +1363,40 @@ describe('cmdheight=0', function()
...
+      {1:~                        }|
+      {1:~                        }|
+      {1:~                        }|
+      {2:[No Name]                }|
+    ]])
+  end)
+
+  it("clears cmdline_area when resized with external messages", function()
+    clear()
+    screen = new_screen({rgb=true, ext_messages=true, ext_linegrid=linegrid})
```
Comment (zeertzjq) [inline test/functional/ui/cmdline_spec.lua:1391]:
```suggestion
    screen = new_screen({rgb=true, ext_messages=true})
```

**Comment by luukvbaal** [comment]
> I'll look into the issue and testcase this weekend.

Did this the same day btw, if it wasn't clear. As far as I'm concerned, latest commit fixes both issues. Unless if someone disagrees with the proposed `'cmdheight'` resizing behavior. I see three possible behaviors:

* Never able to set `'cmdheight'` to zero through resizing.
* Always able to set `'cmdheight'` to zero through resizing.
* Only able to set `'cmdheight'` to zero through resizing if it was explicitly set to zero before.

First commit in this PR implements the third option.

**Pull request merged**